### ISSUE TITLE
[Intern Progress] Munachi Elekwa - Ersilia Outreachy Contribution

### ISSUE THREAD

**Issue by Munaelekwa** [ersilia-os/outreachy-contributions#2]
# 🌟 Intern Progress Tracker  
Welcome! Use this issue to track your progress as you work through the AI/ML modeling exercise. Follow each step, check the boxes when completed, and update your issue as you go.

# ▶️ **Entry tasks**
- [x] Motivation letter posted in this issue
- [x] Installed the Ersilia Model Hub and successfully run models!
- [x] Recorded system settings in this issue
- [x] Explained completed tests of Ersilia in the issue

---

##  **1. Download a Dataset of Interest**
- [x] Created a fork of this repository to work on
- [x] Selected a dataset from Therapeutics Data Commons (TDC)
- [x] Verified that the dataset is a classification problem*
- [x] Read and understood the dataset’s background and endpoint  
- [x] Assessed computational feasibility (size, features, etc.)  
- [x] Downloaded the dataset using Python  
- [x] Saved data in the `/data` folder  
- [x] Started writing documentation in the README (Installation, dataset details, how to run code)  

---

##  **2. Featurise the Data**
- [x] Browsed Ersilia Model Hub for "Representation" models  
- [x] Selected an appropriate featuriser and justified my choice  
- [x] Implemented featurisation in a notebook or script
- [x] Ensured the process is reproducible 
- [x] Updated the README with featurisation instructions  

---

##  **3. Build an ML Model**
- [x] Selected an ML framework
- [x] Trained the model on featurised data  
- [x] Evaluated the model (accuracy, precision, recall, etc.)  
- [x] Created graphs using Matplotlib for better visualization  
- [x] Ensured all steps are reproducible  
- [x] Documented the model-building process in README

---

##  **4. Prepare Code for Review**
- [x] Double-checked that documentation is clear and complete  
- [x] Evaluated if my model performance is satisfactory 
- [x] Identified possible improvements  
- [x] Updated this issue to request a review 

---

##  **5. Stretch Tasks (Optional)**
- [x] Tried a different **featuriser**  
- [x] Tested other **ML architectures**  
- [ ] Applied the model to **public datasets (e.g., ChEMBL)**  
- [ ] Any other improvements? (Describe below)  

---

## 📌 **Requesting Review**
Once you’ve completed **Steps 1-4**, request a review by mentioning a mentor:  
👤 *@GemmaTuron* Please review my work!  

---

### 📝 **Asking questions**
Use this issue to update your progress, explain where you are stuck or ask for review!

Thank you for contributing! 🎉

**Comment by Munaelekwa** [comment]
Hello @GemmaTuron,  

I'm Munachi Elekwa, a software developer and pharmacist, and I’m really excited about the opportunity to contribute to this project. 
My major tech stack is python, django , Flask, Javascript. I have worked on various software development and data science projects using these technologies and I also have a blog on hashnode, where I talk about programming with python and python related topics.

Earlier this year, I set a goal to learn and master AI/ML and get a job with a company that provides AI solutions for healthcare problems. I started out by taking modules of the CS50 Harvard Introduction to Artificial Intelligence with Python course and I have been looking out for opportunities to practice and solidify what I have learnt so far.

While going through the project list for outreachy, I was so happy when I read through the project brief for Ersilia, because it made so much sense for my goals this year and it brings together my drug discovery knowledge from my pharmacy degree and also my programming experience. Contributing to this project is a learning opportunity for me and a good step to achieving my goal of getting a job in Health Informatics. Even if I don't get selected for the internship, I would love to contribute to the project and learn as a volunteer.

I love that this project caters to researchers working in low resource settings and I share in the passion of the project.
I also appreciate learning among a community of people with similar interests and access to mentorship and feedback. I am committed to making the most of this opportunity and contributing significantly to the continous functioning of the project. Thank you.

**Warm regards,**  
Munachi Elekwa

**Comment by Munaelekwa** [comment]
I have successfully installed the *Ersilia Model Hub* and ran the models.
I followed these steps for each model as instructed:

1. ersilia fetch [model id: eos3b5e]
2. ersilia serve [model id]
3. ersilia run -i ibuprofen_input.csv

ps: I saved the smiles notation for ibuprofen, which was given as the test input in a csv file for easy access.
These steps worked for all the three models, here are screen shots of the results:

- eos3b5e

![Image](https://github.com/user-attachments/assets/3cf589fe-ca06-4574-b633-9f08c6d4b0ee)

- eos2r5a

![Image](https://github.com/user-attachments/assets/3de80a3c-9723-42ad-9e25-9858d8301877)

- eos9ei3

![Image](https://github.com/user-attachments/assets/4a390651-e9c7-4efb-a321-9be92f199d12)

**Comment by Munaelekwa** [comment]
**My System Settings**

- OS: Ubuntu 22.04.4 LTS (WSL)
- Python Version: 3.12.9
- Conda Version: 25.1.1
- Docker Version: 27.5.1 

![Image](https://github.com/user-attachments/assets/a746ba56-a59c-41e4-aeb7-3ea075059978)

![Image](https://github.com/user-attachments/assets/bfbc77ae-912d-4c06-a385-a971a5ebc087)

**Comment by Munaelekwa** [comment]
## Explaining the Completed Tests

Firstly, I set up the Ersilia model Hub project  as instructed, by installing the prerequisites : Gcc compiler, miniconda, Docker, and installed ersilia in a new conda environment.
I tried to fetch the models, but it failed because my docker settings was not enabled for wsl integration. I enabled it and tried to fetch the model again, it ran successfully this time.

When i served the first model(eos3b5e) and ran it with the smiles notation for Ibuprofen as input, it returned the molecular weight of Ibuprofen

![Image](https://github.com/user-attachments/assets/3d8807e4-99fe-4aac-b5a1-8c5ac60ad856)

The second model(eos2r5a) returned the retrosynthetic accessibility score for Ibuprofen. This score estimates how easy it is to break down a complex molecule into simpler precursors and it usually falls between 0 and I, with difficulty increasing from 0 to 1. From the estimate given by the model, Ibuprofen has a high RSA - 0.9899, meaning that  it is difficult to break down to smaller precursors.

![Image](https://github.com/user-attachments/assets/c919bfd3-ddfe-4116-8164-39aaabe963db)

The third model(eos9ei3) returned the synthetic accessibility score for Ibuprofen. This score estimates how easy or difficult it is to synthesize a compound from available precursors. The score usually ranges from 1 to 10 with the difficulty increasing from 1 to 10. From the model estimate, Ibuprofen had a synthetic accessibility score of 2.191, implying that it is relatively easy to synthesize Ibuprofen.

![Image](https://github.com/user-attachments/assets/1390fbe2-416d-4167-8625-9899a240e9a8)

**Comment by Munaelekwa** [comment]
@GemmaTuron please review my work!

**Comment by GemmaTuron** [comment]
Hello @Munaelekwa !

Thanks for this good start to the contribution period, great to hear about your interest in AI/ML and our work  Please go ahead and register your first contribution for Ersilia in the Outreachy page linking this issue, where we will continue working during the rest of the internship. If you can, join in the welcome call today at 17:00 CET for more info!

**Comment by Munaelekwa** [comment]
Thank you for your review @GemmaTuron , I will register my contribution right away. Also looking forward to that welcome call later today!

**Comment by Munaelekwa** [comment]
## Download a dataset of Interest

I have created a fork of the repository and selected a dataset from the Therapeutic Data Commons. I have selected the Blood Brain Barrier Permeability dataset. I'm very familiar with ADME processes, so I organically gravitated towards data sets in that category. The blood brain barrier is a membrane that protects the brain and central nervous system and it is highly selective of molecules that penetrate it and this makes the BBB permeability of molecules an important feature in the discovery and formulation of drugs that act on the central nervous system.

I chose the blood brain barrier permeability specifically because the data set is relatively smaller , (1975 drugs) and is suitable for my system's capacity. The data set is structured as a binary classification problem and classification problems are easier and straightforward to evaluate because they have clear endpoints. 

Usually, the blood brain barrier permeability is determined by the value of the parameter- Log BB. This is determined experimentally by measuring the concentration of the drug in the blood and the concentration of the drug in the brain and calculating log of the ratio of the drug's concentration in the brain to it's concentration in the blood. 

$$
Log BB = Log\frac{Cbrain}{Cblood}
$$

Molecules are classified as permeable by the BBB(BBB+) if $Log BB \geq -1$ and impermeable (BBB-) if Log BB < -1. This dataset labels BBB+ as 1 and BBB- as 0.

The model will take in the smiles notation of the drug/molecule as input and predict it's ability to permeate the BBB.

I have downloaded the dataset with python and saved the data in the data folder.

**Comment by GemmaTuron** [comment]
Good start @Munaelekwa 

In classification problems it is important to specify how the problem was converted to binary (what cutoff, what was exactly measured experimentally etc)- can you add this information and start building the readme file in your fork?

**Comment by Munaelekwa** [comment]
> Good start [@Munaelekwa](https://github.com/Munaelekwa)
> 
> In classification problems it is important to specify how the problem was converted to binary (what cutoff, what was exactly measured experimentally etc)- can you add this information and start building the readme file in your fork?

Thank You @GemmaTuron  for the feedback, i have updated the comment with this information, and I have also started building my readme documentation, (I will open a PR to this effect). let me know if there's any information missing.

**Comment by Munaelekwa** [comment]
## Featurising the Data

I explored the models tagged 'representation' in ersilia model hub and found a couple of useful featurisers for my project data. I was looking out for models that represent rich structural and physicochemical properties of compounds, since blood brain permeability is well influenced by molecular weight, lipophilicity, polarity etc. The morgan fingerprints and rd_kit descriptors looked promising to me, so i did a little research to help me make a decision and I chose the rd_kit decsriptor because according to my research, it provides richer feature set for machine learning. I would still explore the morgan finger prints later, and compare the model efficiency.

After choosing the featuriser, I wrote a python script to batch featurise the data and it kept failing until i asked for help on the slack channel and a fellow contributor pointed me to the method described in the ersilia documentation. This has taught me that I should read and consult documentation more often );
I modified the script with the method and it worked. I have also saved the featurised data for all splits (train, test and valid) of my dataset in the data folder.

Next on my to do is to update my README file with the featurisation instructions and start building my model.

**Comment by Munaelekwa** [comment]
Update: The featurisation instructions have now been updated in the README file

**Comment by GemmaTuron** [comment]
Hi @Munaelekwa 

Thanks for the summary, yes, always check documentation whenever possible. The RDKit descriptors is a fine choice for a baseline model indeed. I have looked at the open PR and it looks good to me, please go ahead with next steps!

**Comment by Munaelekwa** [comment]
I have started on the model building. I selected the Scikit-learn framework because I have mostly used it in my learning. I had to modify my featurised data, because I noticed the featurised data excludes the Y column(permeability) after generating the numerical representation of the molecules. I modified the featurisation script to include the target column after featurisation and ran the script again.

After getting the updated featurised data, i trained the Random Forest Model with the featurised train data and evaluated the prediction performance with the featurised valid data. Here's a screenshot of the classification report: 

![Image](https://github.com/user-attachments/assets/fdda33ee-a27d-4c78-bbba-325101566ad1)

from the report I'm seeing that the model has an accuracy of 89.16% and generally predicts BBB+(recall=0.97) more accurately than it does BBB-(recall=0.65) and it might be because the data has more BBB+ves, (152) than BBB-ves(51)

I'm thinking of oversampling with SMOTE, to increase the BBB- class for better performance, and also exploring the XGBoost framework.

**Comment by Munaelekwa** [comment]
Hi @GemmaTuron,

I have completed the model building process, and  my model, a random forest classifier trained on an rdkit descriptor featurised data achieved an AUROC score of **0.9144**, which i consider well performing, considering that it places 4th on the current [TDC leaderboard](https://tdcommons.ai/benchmark/admet_group/BBB_Martins) for this dataset, with the highest performing model reaching an Auroc score of 0.92. I am also exploring a couple of methods that could make the performance better, but so far, this works.

Here's a summary of my model performance: 
### Model Performance Summary  

| **Metric**   | **Train**  | **Validation** | **Test**  |
|-------------|-----------|--------------|---------|
| **Accuracy**  | 99.79%    | 89.16%       | 86.70%  |
| **Precision** | 100.00%    | 89.00%       | 86.00%  |
| **Recall**    | 100.00%    | 89.00%       | 87.00%  |
| **F1-Score**  | 100.00%    | 89.00%       | 86.00%  |
| **AUROC**   | 100.00%    | 91.72%       | 91.44%  |

I also wrote a comprehensive jupyter notebook of my model evaluation which includes the ROC curve, confusion matrix, precision-recall curve and a feature importance plot that shows the top 10 features that helped the model's decision making. 
These updates are already committed and can be seen in my open PR.

Next on my to-do is to update my readme file with the model building process and finalize my documentation so that I can ask for an official review, while I go ahead to explore the stretch tasks.

What do you think so far?

**Comment by GemmaTuron** [comment]
Thanks @Munaelekwa 

This is looking good! Can you save the graphs in a /figures folder so I can see them while you finish the documentation of the project?

**Comment by Munaelekwa** [comment]
> Thanks [@Munaelekwa](https://github.com/Munaelekwa)
> 
> This is looking good! Can you save the graphs in a /figures folder so I can see them while you finish the documentation of the project?

Thank you, @GemmaTuron , I have saved the analysis graphs in the data/figures folder and committed to PR

**Comment by Munaelekwa** [comment]
My documentation is now updated with model building process and evaluation. Just going through my work and preparing my code for review

**Comment by GemmaTuron** [comment]
Perfect thanks! When you complete this please drop a summary here with the main takings @Munaelekwa for review! Thanks

**Comment by Munaelekwa** [comment]
## Summary of Work Done So Far
This project has been exciting to work on, and also it has been a learning process. I'm happy I picked this particular dataset.
Here's a summary of what I have done so far:

- I selected the blood brain barrier permeability dataset from TDC
- I featurised the dataset using the rd_kit physicochemical descriptors from ersilia model hub.
- I trained a random forest classifier model from Scikit-learn using the featurised data
- The model on evaluation performed well, with an auroc score of 0.9144, ranking 4th on the Tdc leaderboard for this dataset.
- The model can be improved on, as the confusion matrix reveals a significant number of false positives( class 0 compounds misclassified as class 1). Indicating that the model struggles a bit with specificity, and this could be due to the class imbalance of the dataset. Class 1 is overrepresented (1551: 479), leaving the model with little to work with for class 0.
- So far, training an xgboost model with the same dataset or oversampling data with SMOTE did not yield an improvement in model performance. I want to explore using morgan fingerprints as a featuriser to see the impact on model performance, and play around with training parameters to see if there would be an improvement. (more updates on this as a stretch task)
- I also deployed this project as a streamlit app, with a mini UI to enable non-technical users and everyone generally to try out the model. You can test it at https://bbbpredictor.streamlit.app/.
- I have reviewed my code and documentation and I think I need a review at this point to point out any areas to be expanded on or adjusted.
Thank you for your continued guidance @GemmaTuron , I would appreciate your feedback 😊

**Comment by Munaelekwa** [comment]
## Stretch Tasks

I explored another featuriser, the morgan fingerprints from ersilia model hub and also tried using xgboost as model framework, with the following settings: 

```bash
xgb_model = xgb.XGBClassifier(
    objective="binary:logistic",
    eval_metric="logloss",
    use_label_encoder=False,
    learning_rate=0.1,
    n_estimators=100,
    max_depth=6,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42
)
```

Here is the overall summary for the featuriser-model combinations explored:

| Model                                                  | Precision    | Recall | Accuracy |   F1 Score |    AUROC |
|------------------------------------------|-------------|--------|-----------|-------------|-----------|
| RDKit + Random Forest                       | 0.86       |    0.87       |    0.87            |        0.86           |      0.9144          |
| RDKit + Random Forest + SMOTE       |     0.85      |  0.80         |    0.87            |   0.73        |   0.9047             |
| RDKit + XGBoost                                  |     0.83              |     0.77      |     0.86           |       0.80            |       0.8969         |
| Morgan Fingerprint + Random Forest |        0.83           |    0.70       |     0.83           |        0.73           |      0.8394          |
| Morgan Fingerprint + RF + SMOTE     |     0.85               |   0.74       |      0.85          |       0.77            |      0.8522          |
| Morgan Fingerprint + XGBoost           |     0.83               |     0.71      |     0.83          |     0.74              |      0.8829           |

from the results, the rdkit-featurised data trained random forest model remains the best performing model for this task.

**Comment by GemmaTuron** [comment]
Hi @Munaelekwa !!

I love the streamlit app many thanks for setting this up, nice job, I tried it out and works well! The summary is also helpful. Please make sure that the code and documentation on your repo are completed for the final review. I'll now add some more instructions via Slack !

**Comment by Munaelekwa** [comment]
Here's a link to my fork of the project - https://github.com/Munaelekwa/outreachy-contributions/tree/Munachi-Contribution and PR #35

**Comment by GemmaTuron** [comment]
Hi @Munaelekwa 

Many thanks for your work these days, great to see your advances. Please start preparing your final application as set out in Slack, and remember to link here any model refactoring attempts you might have contributed to!

**Comment by Munaelekwa** [comment]
Thank you @GemmaTuron , I have started preparing my final application. 

Some of the models I tried to fetch to contribute to the model refactoring returned errors and I opened issues to this effect. Does this count as a valid contribution?

**Comment by GemmaTuron** [comment]
Hi @Munaelekwa 

Yes please do add mention them here, I am looking at the models these days to see the errors!

**Comment by Munaelekwa** [comment]
Here are some of my Contributions to Model Refactoring @GemmaTuron:
I was getting a bentoML version error, so i opened these issues:

[eos9c7k](https://github.com/ersilia-os/eos9c7k/issues/10)
[eos4b8j](https://github.com/ersilia-os/eos4b8j/issues/6)
[eos9be7](https://github.com/ersilia-os/eos9be7/issues/5)

I also tried to fetch these models: `eos0t04`, `eos526j`, `eos85a3`, `eos8451`, `eos6fza`, `eos1vms`, `eos92sw`
- I couldn't find `eos0t04` in repositories page
- I was able to fetch and run `eos526j` and generate a `run_input` file, but it returned a `Type error: Nonetype object is not iterable` when trying to copy output to a ``run_output` file
- The other models returned the bentoml error and I tried to use the alternative method you suggested on slack `run.sh`, it seems to be taking some time and eventually disconnects wsl. I have fixed this and would try again when I have completed my final application, if it has not been done yet.

Thank you for your support and encouragement this contribution period.

**Comment by GemmaTuron** [comment]
Hi @Munaelekwa 

Thanks for the summary!

**Comment by miquelduranfrigola** [comment]
Auto-closing this Outreachy intern-progress tracker as completed — the contribution round has ended. Reopen if follow-up is needed.
